File: rdos-validation-tools/rdo-validation-tool-scala/src/main/scala/com/rms/rdo/validation/datatype/DataTypeValidator.scala
package com.rms.rdo.validation.datatype

import com.rms.rdo.validation.constant.Constants
import com.rms.rdo.validation.model.DataTypeValidationError

import scala.util.Try

object DataTypeValidator {

  def validateDataType(row: Map[String, String],
                       validationMap: Map[String, String]): List[DataTypeValidationError] = {
    val columnNames = row.keySet.toList
    columnNames.flatMap { columnName =>
      checkDataTypes(row(columnName), validationMap(columnName))
        .map(error => DataTypeValidationError(columnName, row(columnName), error))
    }
  }

  private def checkDataTypes(columnValue: String, dataType: String): Option[String] = {
    dataType match {
      case "Date" => Validator.validateDateAndTimeDataType(columnValue, Constants.DateFormats, dataType)
      case "Time" => Validator.validateDateAndTimeDataType(columnValue, Constants.TimeFormats, dataType)
      case "DateTime" => Validator.validateDateAndTimeDataType(columnValue,
        Constants.DateTimeFormats, dataType)
      case "Bit" => Validator.validateBitType(columnValue, dataType)
      case _ => validateNumericDataTypes(columnValue, dataType)
    }
  }

  private def validateNumericDataTypes(columnValue: String, dataType: String): Option[String] = {
    dataType match {
      case "Double" => Validator.validateNumericType(Try(columnValue.toDouble).toOption, dataType)
      case "Long" => Validator.validateNumericType(Try(columnValue.toLong).toOption, dataType)
      case "Byte" => Validator.validateNumericType(Try(columnValue.toByte).toOption, dataType)
      case "Short" => Validator.validateNumericType(Try(columnValue.toShort).toOption, dataType)
      case "Integer" => Validator.validateNumericType(Try(columnValue.toInt).toOption, dataType)
      case _ => None
    }
  }
}

File: rdos-validation-tools/rdo-validation-tool-java/src/main/java/com/rms/rdo/validation/datatype/DataTypeValidator.java
package com.rms.rdo.validation.datatype;

import com.rms.rdo.validation.constants.Constants;
import com.rms.rdo.validation.model.datatype.DataTypeValidationError;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class DataTypeValidator {

    public static List<DataTypeValidationError> validateDataType(Map<String, String> row,
                                                                 Map<String, String> validationMap) {
        List<String> columnNames = new ArrayList<>(row.keySet());
        return columnNames.stream().map(columnName ->
                checkDataTypes(row.get(columnName), validationMap.get(columnName))
                        .map(error -> DataTypeValidationError.builder()
                                .columnName(columnName)
                                .columnValue(row.get(columnName))
                                .errorMessage(error)
                                .build()))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    private static Optional<String> checkDataTypes(String columnValue, String dataType) {

        switch (dataType) {
            case "Double":
                return Validator.validateNumericType(() -> Double.valueOf(columnValue).toString(), dataType);
            case "Long":
                return Validator.validateNumericType(() -> Long.valueOf(columnValue).toString(), dataType);
            case "Byte":
                Validator.validateNumericType(() -> Byte.valueOf(columnValue).toString(), dataType);
            case "Short":
                return Validator.validateNumericType(() -> Short.valueOf(columnValue).toString(), dataType);
            case "Integer":
                return Validator.validateNumericType(() -> Integer.valueOf(columnValue).toString(), dataType);
            case "Date":
                return Validator.validateDateAndTimeDataType(columnValue, Constants.DATE_FORMATS, dataType);
            case "Time":
                return Validator.validateDateAndTimeDataType(columnValue, Constants.TIME_FORMATS, dataType);
            case "DateTime":
                return Validator.validateDateAndTimeDataType(columnValue, Constants.DATE_TIME_FORMATS, dataType);
            case "Bit":
                return Validator.validateBitType(columnValue, dataType);
            default:
                return Optional.empty();
        }
    }
}
